File: data-structure-and-algorithm/typescript/src/core/queue/PriorityQueue.ts
import IQueue from "./IQueue";
import BinaryHeap from "../heap/BinaryHeap";
export default class PriorityQueue<T> implements IQueue<T> {
  private _heap: BinaryHeap<T>;
  constructor(comparator: (t1: T, t2: T) => number) {
    this._heap = new BinaryHeap<T>(comparator);
  }
  size(): number {
    return this._heap.size();
  }
  isEmpty(): boolean {
    return this._heap.isEmpty();
  }
  clear(): void {
    this._heap.clear();
  }
  enQueue(element: T): void {
    this._heap.add(element);
  }
  deQueue(): T {
    return this._heap.remove();
  }
  front(): T {
    return this._heap.get();
  }
}

File: data-structure-and-algorithm/java/src/com/bestlyg/queue/PriorityQueue.java
package com.bestlyg.queue;

import java.util.Comparator;

import com.bestlyg.heap.BinaryHeap;

public class PriorityQueue<E> {
	private BinaryHeap<E> heap = new BinaryHeap<>();

	public PriorityQueue(Comparator<E> comparator) {
		heap = new BinaryHeap<>(comparator);
	}

	public PriorityQueue() {
		this(null);
	}

	public int size() {
		return heap.size();
	}

	public boolean isEmpty() {
		return heap.isEmpty();
	}

	public void clear() {
		heap.clear();
	}

	public void enQueue(E element) {
		heap.add(element);
	}

	public E deQueue() {
		return heap.remove();
	}

	public E front() {
		return heap.get();
	}
}
